### PR TITLE
ENH: Allow expected hessian in scipy optimizer fits

### PR DESCRIPTION
**Pull request by thequackdaddy**
Hello,

I think this is a rather simple approach to exposing the expected Hessian for scipy optimizer fits. 

I thought about using functools `partial` but ran into some obstacles. The real issue is that context matters and not all the `hessian` methods have an `observed` parameter. So this was as simple as I could make it without much effort. 

xref #3257

### CODE REVIEW

File: statsmodels/base/model.py [hunk 1]
```diff
@@ -436,14 +436,17 @@ def fit(self, start_params=None, method='newton', maxiter=100,
         nobs = self.endog.shape[0]
         f = lambda params, *args: -self.loglike(params, *args) / nobs
         score = lambda params, *args: -self.score(params, *args) / nobs
-        try:
+        if observed_hessian:
             hess = lambda params, *args: -self.hessian(params, *args) / nobs
-        except:
-            hess = None
+        else:
+            hess = lambda params, *args: -self.hessian(params, *args, observed=False) / nobs
```
Comment (josef-pkt) [inline statsmodels/base/model.py]: nice trick
I was wondering why this doesn't break things in the standard case where observed is not an available keyword.
But you don't change the code on the default path.

I don't know why we had the try with catch-all except.

**Comment by josef-pkt** [comment]
statespace models which are the only ones that currently have a similar option use `optim_hessian` with 3 choices http://www.statsmodels.org/devel/generated/statsmodels.tsa.statespace.sarimax.SARIMAX.fit.html

(IIRC or if I guess correctly, then OPG and EIM are the same in GLM, at least for fixed scale/dispersion)

**Comment by thequackdaddy** [comment]
As far as I could tell at runtime, lamdas only have their parameters evaluated. Not until the lambda is called will the function fully evaluate. Additionally, the variables that are used must in the scope of the calling module. The nobs divisor needs to be declared in the calling module and you have to physically pass "True"

Honestly, this really feels like something that you should use functools partial... but the divide by nobs part and using negative for all optimizers but newton made that do-able.

File: statsmodels/base/model.py [hunk 2]
```diff
@@ -253,7 +253,7 @@ def hessian(self, params):
 
     def fit(self, start_params=None, method='newton', maxiter=100,
             full_output=True, disp=True, fargs=(), callback=None, retall=False,
-            skip_hessian=False, **kwargs):
+            skip_hessian=False, observed_hessian=True, **kwargs):
```
Comment (josef-pkt) [inline statsmodels/base/model.py]: the new keyword here is not used anymore after the last commit

**Comment by josef-pkt** [comment]
We might have to change observed=None, because hessian would ignore the observed keyword when called directly and _optim_hessian is set, which would be a difficult source of bugs to find.

File: statsmodels/genmod/generalized_linear_model.py [hunk 3]
```diff
@@ -605,6 +604,8 @@ def hessian(self, params, scale=None, observed=True):
         hessian : ndarray
             Hessian, i.e. observed information, or expected information matrix.
         """
+        if hasattr(self, '_optim_hessian') and self._optim_hessian == 'eim':
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: switch to `observed=None` in signature.
and protect this with `if observed is None:   ....  else observed=True`

**Comment by josef-pkt** [comment]
What happened to the observed=None in`hessian(self, params, scale=None, observed=True)` which I thought I saw in an earlier version of the last commit?  Was there a problem with it?

**Comment by thequackdaddy** [comment]
It still [there](https://github.com/thequackdaddy/statsmodels/blob/9b02c1237f85c19355f9081e68e2939eab7eec8a/statsmodels/genmod/generalized_linear_model.py#L587).

I amended the commit because I had logic that essentially said if `observed` is not `None`, the `observed = True`... so when you tried to compare to the EIM during non-fits, you got the OIM instead.

**Comment by josef-pkt** [comment]
Good, github changes view showed the old signature.

merging, Thanks @thequackdaddy

**Pull request merged**